miasm2/core/objc.py

```diff
@@ -0,0 +1,1712 @@
+        subobj = ast_get_align_size(ast.type, types_mngr)
+        dim = ast.dim
+        value = ast_eval_int(dim, types_mngr)
+        obj = ObjCArray(subobj, value)
+    elif isinstance(ast, c_ast.Union):
+        obj = ast_gen_union_align_size(ast, types_mngr)
+    elif isinstance(ast, c_ast.Struct):
+        obj = ast_gen_struct_align_size(ast, types_mngr)
+    elif isinstance(ast, c_ast.PtrDecl):
+        void_type = types_mngr.void_ptr
+        subobj = ast_get_align_size(ast.type, types_mngr)
+        obj = ObjCPtr('noname', subobj, void_type.align, void_type.size)
+    else:
+        raise NotImplementedError("Not implemented!")
+    assert isinstance(obj, _ObjCRecurse) or obj.align in [
+        1, 2, 4, 8, 16, 32, 64, 128, 256]
+    return obj
+
+
+def ast_gen_struct_align_size_compil(ast, types_mngr):
+    """Evaluates the size/alignment of a C ast structure
+    (default packed)
+
+    @ast: parsed pycparser.c_ast object
+    @types_mngr: types manager
+    """
+
+    offset = 0
+    align_max = 1
+
+    if ast.name is None:
+        name = types_mngr.gen_temp_name()
+    else:
+        name = ast.name
+    new_obj = ObjCStruct(name)
+    if ast.decls is None:
+        # If object is unknown, it's a recursive struct
+        if types_mngr.is_known_type((name,)):
+            obj = types_mngr.get_type((name,))
+        else:
+            obj = _ObjCRecurse(name)
+        return obj
+    for arg in ast.decls:
+        obj = ast_get_align_size(arg, types_mngr)
+        align_max = max(align_max, obj.align)
+        if obj.align > 1:
+            offset = (offset + obj.align - 1) & ~(obj.align - 1)
+        new_obj.add_field(arg.name, obj,
+                          offset, obj.size)
+        offset += obj.size
+
+    # Structure alignement is its field max alignement
+    align = align_max
+    if align > 1:
+        offset = (offset + align - 1) & ~(align - 1)
+
+    new_obj.set_align_size(align, offset)
+
+    types_mngr.add_type((name, ), new_obj)
+    return new_obj
+
+
+def ast_gen_struct_align_size_packed(ast, types_mngr):
+    """Evaluates the size/alignment of a C ast structure
+    (packed)
+
+    @ast: parsed pycparser.c_ast object
+    @types_mngr: types manager
+    """
+
+    offset = 0
+
+    if ast.name is None:
+        name = types_mngr.gen_temp_name()
+    else:
+        name = ast.name
+    new_obj = ObjCStruct(name)
+    if ast.decls is None:
+        # If object is unknown, it's a recursive struct
+        if types_mngr.is_known_type((name,)):
+            obj = types_mngr.get_type((name,))
+        else:
+            obj = _ObjCRecurse(name)
+        return obj
+
+    for arg in ast.decls:
+        obj = ast_get_align_size(arg, types_mngr)
+        new_obj.add_field(arg.name, obj,
+                          offset, obj.size)
+        offset += obj.size
+
+    new_obj.set_align_size(1, offset)
+
+    types_mngr.add_type((name, ), new_obj)
+    return new_obj
+
+
+def ast_gen_struct_align_size(ast, types_mngr):
```

Comment:
Note for future: to avoid having a lot of module method, and for a better "management" and readability of layout, it could be great to have a `Layout` class, describing how a given layout act on union, align size, ...